### PR TITLE
[WIP] Added CRD controller

### LINKED ISSUE #13: Add CRD to create secrets wrapper
Many existing applications running on Kubernetes relays on secrets to store sensitive data - which make Kamus unusable. This can be solved with a CRD that is wrapper around secret, something in the following format:
```
apiVersion: v1
kind: KamusSecret
metadata:
  name: mysecret
  namespace: default
  service-account: default
type: Opaque
data:
  username: <encrypted data>
  password: <encrypted data>
```
The CRD controller will use Kamus and the SA token to decrypt the value of the secret and create a real Kubernetes secret object from it.

### PR DESCRIPTION
**Pull request by omerlh**
This is a very first MVP version of the CRD. Currently, only create and delete are supported. 
In order to finish this task we need:
 - [x] Documentation
 - [x] Tests
 - [x] Release a new docker image
 - [x] Run the tests in the CI/CD + make sure all tests are passing
 - [x] Test using multiple k8s versions
 - [x] Add regression tests - encrypt data using the code on master and test it's decrypted succesfully. 

Want to give it a try? You can build the images locally, and use `crd.yaml` and `tls.yaml` to test it out.

close #13

### CODE REVIEW

**Comment by omerlh** [comment]
To test the CRD:
* Clone the code, and make sure you have a cluster running (use `kubectl get pods` to check).
* `cd src/crd-controller`
* `kubectl apply -f crd.yaml `
* Now run `dotnet run` and wait to see 
```
[<time> INF] CRD controller started
```
* Run `kubectl apply -f tls.yaml ` from another terminal. This will create a CRD object. Now run `kubectl get secrets` and look for a secret named `my-tls-secret`. It's working!
* Run `kubectl delete -f tls.yaml ` to delete the CRD object. Notice that `kubectl get secrets`  does not return a secret named `my-tls-secret`. It's working!

File: src/crd-controller/HealthChecks/KubernetesPermissionsHelthCheck.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Serilog;
+
+namespace CustomResourceDescriptorController.HealthChecks
+{
+    public class KubernetesPermissionsHelthCheck : IHealthCheck
+    {
+        private readonly IKubernetes mKubernetes;
+        private readonly ILogger mLogger = Log.ForContext<KubernetesPermissionsHelthCheck>();
+        private readonly MemoryCache _memoryCache = new MemoryCache(new MemoryCacheOptions());
```
Comment (shaikatz) [inline src/crd-controller/HealthChecks/KubernetesPermissionsHelthCheck.cs]:
```suggestion
        private static readonly MemoryCache _memoryCache = new MemoryCache(new MemoryCacheOptions());
```

File: src/crd-controller/LoggingMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace CustomResourceDescriptorController
+{
+    // You may need to install the Microsoft.AspNetCore.Http.Abstractions package into your project
```
Comment (shaikatz) [inline src/crd-controller/LoggingMiddleware.cs]:
```suggestion
```

File: src/crd-controller/Models/KamusSecret.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+using k8s.Models;
+
+namespace CustomResourceDescriptorController.Models
+{
+    public class KamusSecret : KubernetesObject
+    {
+        public Dictionary<string, string> Data { get; set; }
+        public string Type { get; set; }
+        public V1ObjectMeta Metadata { get; set; }
+        public string ServiceAccount { get; set; }
```
Comment (shaikatz) [inline src/crd-controller/Models/KamusSecret.cs]: Isn't that an issue where the property names are capitalized here and camel cased in the `KamusSecret` k8s object (in the docs)?

File: src/crd-controller/Program.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
+            WebHost.CreateDefaultBuilder(args)
+                .UseMetrics(options => { 
+                    options.EndpointOptions = endpointsOptions => {
+                        endpointsOptions.MetricsEndpointOutputFormatter = new MetricsPrometheusTextOutputFormatter();
+                    };
+                })
+
```
Comment (shaikatz) [inline src/crd-controller/Program.cs]:
```suggestion
```

File: src/crd-controller/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        public IConfiguration Configuration;
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices (IServiceCollection services) {
+
+            services.AddSingleton(Configuration);
+            services.AddMvc ()
+                    .SetCompatibilityVersion(CompatibilityVersion.Version_2_1);
```
Comment (shaikatz) [inline src/crd-controller/Startup.cs]:
```suggestion
                    .SetCompatibilityVersion();
```

Comment (shaikatz) [reply src/crd-controller/Startup.cs]: I think it's not needed anymore.

File: src/crd-controller/crd-controller.csproj [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    <Folder Include="HostedServices\" />
+    <Folder Include="Models\" />
+    <Folder Include="HealthChecks\" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="KubernetesClient" Version="1.5.1-g60ffd20a7c" />
+    <PackageReference Include="System.Net.Http" Version="4.3.3" />
+    <PackageReference Include="App.Metrics.AspNetCore.Mvc" Version="2.0.0" />
+    <PackageReference Include="App.Metrics.Formatters.Prometheus" Version="2.0.0" />
```
Comment (shaikatz) [inline src/crd-controller/crd-controller.csproj]: Please update both AppMetrics packages to version 3 for newest Prometheus support

File: src/crd-controller/Models/KamusSecret.cs (hunk 3 above)
Comment (omerlh) [reply src/crd-controller/Models/KamusSecret.cs]: It worked, I'm not 100% sure why 🤷‍♂️

File: src/crd-controller/crd-controller.csproj (hunk 6 above)
Comment (omerlh) [reply src/crd-controller/crd-controller.csproj]: I will!

**Comment by omerlh** [comment]
Cool, thanks! Just left to fight the build :)

**Pull request merged**